staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -577,20 +601,41 @@ func (s *store) List(ctx context.Context, key string, opts storage.ListOptions,
 			returnedRV = continueRV
 		}
 	case s.pagingEnabled && pred.Limit > 0:
-		if len(resourceVersion) > 0 {
-			fromRV, err := s.versioner.ParseResourceVersion(resourceVersion)
-			if err != nil {
-				return apierrors.NewBadRequest(fmt.Sprintf("invalid resource version: %v", err))
-			}
-			if fromRV > 0 {
-				options = append(options, clientv3.WithRev(int64(fromRV)))
+		if fromRV != nil {
+			switch match {
+			case metav1.ResourceVersionMatchNotOlderThan: // do nothing
+			case metav1.ResourceVersionMatchExact:
+				returnedRV = int64(*fromRV)
+				if returnedRV == 0 {
+					// Special case: exact + rv=0. etcd keyspace is empty. xref https://github.com/etcd-io/etcd/issues/11496
+					return s.versioner.UpdateList(listObj, 0, "", nil)
```

Comment:
this is ... interesting ... won't this mess up a client that tries to start a watch from the returned listmeta resourceVersion? what does starting a watch with `?watch=true&resourceVersion=0&resourceVersionMatch=Exact` do?

Reply:
For watch `resourceVersionMatch` is ignored, right?

Reply:
I'll add a validation rule forbidding resourceVersionMatch to be set for watch since it is only respected/needed for list calls.

Reply:
I'm still unsure about returning lists with resourceVersion=0, given how we interpret subsequent watch requests a client could reasonably make treating that as an opaque value. Would it be better to prevent this combination in validation?

Reply:
I'm okay with preventing it in validation. It makes everything else simpler. I'll make the changes and add back the assertion that checks for rv=0 in responses.